pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1526,6 +1528,23 @@ func (c *PodAffinityChecker) satisfiesPodsAffinityAntiAffinity(pod *v1.Pod,
 	return nil, nil
 }
 
+// CheckNodeLostPredicate makes sure a pod won't be scheduled to a NodeLost node
+func CheckNodeLostPredicate(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	if nodeInfo == nil || nodeInfo.Node() == nil {
+		return false, []algorithm.PredicateFailureReason{ErrNodeUnknownCondition}, nil
+	}
+	// sometimes worker nodes can't communicate back to apisever due to network partition
+	// then no networkUnschedulable taint is not received, so PodToleratesNodeTaints won't help
+	for _, cond := range nodeInfo.Node().Status.Conditions {
+		// using following logic to determine semantics of "NodeLost"
+		if cond.Type == v1.NodeReady && cond.Status == v1.ConditionUnknown &&
+			cond.Message == "Kubelet stopped posting node status." {
```

Comment:
Put it here for brainstorming - is this an excessive check?

If not, I will put the string to a const variable.

Reply:
> Relying on the exact message looks very flaky to me.

yeap, not sure there is a better checking criteria to harden the logic here. 

> The original issue says that the pods on a node with lost status are not evicted from the node

My reading (through the autoscaler e2e code) is that, they put a pod onto a healthy node first, and then ssh to that node to break the network connectivity (through iptables command). Then check if autoscaler is working fine to auto scale another replacement node. Finally kickoff a series of mem pods and check whether they are scheduled well.

@aleksandra-malinowska could you confirm the e2e logic? ^^

BTW: I did a test disabling TaintNodesByCondition:

- Create a 1-node cluster
- Create a deployment with 10 replicas, and they landed on the only node with running status
- SSH into the node and break connectivity to apiserver
- Wait for the node-1 to be NotReady (not that at this moment, the 10 pods are still running status, but internally the Ready condition is false, I guess it's apiserver modifying them when it's not getting heartbeat from kubelet within a time period)
- Then create node-2 and join the cluster
- 8 mins later (interesting time), another 10 pods were created with running status onto node-2. And old 10 pods switches to Unknown status.

If I enable TaintNodesByCondition, result of last step above is:
- 8 mins later (interesting time), 5 pods were created with running status onto node-2, and 5 pods still get scheduled to node-1 with Pending status. And old 10 pods switches to Unknown status.

Reply:
@Huang-Wei when TaintNodeByCondition is enabled, is the node with unavailable network tainted properly? This is something that node_lifecycle_controller should do.

Reply:
> when TaintNodeByCondition is enabled, is the node with unavailable network tainted properly?

nope, there is no taint - actually the "unavailable network" condition/taint is always confusing to me. If the network gets broken totally, how kubelet reports back to apiserver... I guess the explicit "unavailable network" condition/taint is more a "network not configured well" case, say the connectivity works, it's just network provider (calico) not configuring/working well.

> This is something that node_lifecycle_controller should do.

Not sure it's node_lifecycle_controller or kubelet.
- If it's node_lifecycle_controller, then I believe there is bug fix should be given there.
- If it's kubelet, I believe it's trying to report taint back. But due to network partition, they can't report the status back to apiserver...

Reply:
Sounds good to me.

Reply:
Did not check the log; but I think we should taint NoSchedule for unreachable. If network broken, node_controller will update node's condition to `Unreachable`. node_controller will evict the pod if node unreachable, if node disappeard, pod gc will delete the pod. We need to make sure the new pod will not be scheduled to the node if object in API is still there. Regarding `Network Unavailable` may update by network provider which means the configuration is incorrect.

Reply:
Thanks @k82cn @bsalamat @ravisantoshgudimetla for your reviewing.

I think the consensus is to fix this issue in node lifecycle controller side. Specifically, to add a "unreachable" taint with "NoSchedule" effect when the node lost connection with apiserver. I will create another PR and close this one.

 Appreciating your comments again.

Reply:
Example run of failing test: https://k8s-testgrid.appspot.com/sig-autoscaling-cluster-autoscaler#gci-gce-autoscaling
Example failing pod: memory-reservation-45ntl
kube-controller-manager log: https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling/7156/artifacts/ca-master/kube-controller-manager.log

Searching for one such pod in kube-controller-manager log, it looks like it tries to delete it:

```I0822 02:06:38.728725       1 controller_utils.go:212] Recording Deleting all Pods from Node ca-minion-group-ppzt. event message for node ca-minion-group-ppzt
I0822 02:06:38.736220       1 replica_set.go:338] Pod memory-reservation-45ntl updated, objectMeta {Name:memory-reservation-45ntl GenerateName:memory-reservation- Namespace:e2e-tests-autoscaling-pk9fb SelfLink:/api/v1/namespaces/e2e-tests-autoscaling-pk9fb/pods/memory-reservation-45ntl UID:652631ca-a5af-11e8-8b55-42010a800002 ResourceVersion:29348 Generation:0 CreationTimestamp:2018-08-22 02:02:16 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[name:memory-reservation] Annotations:map[] OwnerReferences:[{APIVersion:v1 Kind:ReplicationController Name:memory-reservation UID:6498b825-a5af-11e8-8b55-42010a800002 Controller:0xc4239ecafe BlockOwnerDeletion:0xc4239ecaff}] Initializers:nil Finalizers:[] ClusterName:} -> {Name:memory-reservation-45ntl GenerateName:memory-reservation- Namespace:e2e-tests-autoscaling-pk9fb SelfLink:/api/v1/namespaces/e2e-tests-autoscaling-pk9fb/pods/memory-reservation-45ntl UID:652631ca-a5af-11e8-8b55-42010a800002 ResourceVersion:30380 Generation:0 CreationTimestamp:2018-08-22 02:02:16 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[name:memory-reservation] Annotations:map[] OwnerReferences:[{APIVersion:v1 Kind:ReplicationController Name:memory-reservation UID:6498b825-a5af-11e8-8b55-42010a800002 Controller:0xc4235f1a2e BlockOwnerDeletion:0xc4235f1a2f}] Initializers:nil Finalizers:[] ClusterName:}.
I0822 02:06:38.736477       1 controller_utils.go:183] Controller expectations fulfilled &controller.ControlleeExpectations{add:0, del:0, key:"e2e-tests-autoscaling-pk9fb/memory-reservation", timestamp:time.Time{wall:0xbed74eea02ba0c12, ext:8967498278401, loc:(*time.Location)(0x5a9a1a0)}}```

Afterwards, there are still updates for this pod. So perhaps the issue is that the pod isn't successfully deleted?

Reply:
@aleksandra-malinowska hmm... I will leave comments with symptoms described in https://github.com/kubernetes/kubernetes/pull/67677#discussion_r211781390 under your issue. I believe they're related. And then I will create a PR to see if it fixes your issue.